### PR TITLE
perf: Improve count aggregate performance

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion-comet/issues/744

## Rationale for this change

For some reason, `COUNT` is really slow when used from `Comet`, but `SUM` is fast, so let's translate `COUNT(expr)` to `SUM(IF(expr IS NULL, 0, 1))` ~until we can get to the bottom of the real issue~.

edit: It turns out that Spark also implements `COUNT` this way, so I think this closes the issue.

```
Grouped HashAgg Exec: single group key (cardinality 1048576), single aggregate COUNT:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
--------------------------------------------------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark (COUNT)                                                                    1716           1728          17          6.1         163.7       1.0X
SQL Parquet - Comet (Scan) (COUNT)                                                             1677           1680           4          6.3         159.9       1.0X
SQL Parquet - Comet (Scan, Exec) (COUNT)                                                        782            800          27         13.4          74.6       2.2X
```

### CODE REVIEW

**Comment by andygrove** [comment]
This PR results in an 11% speedup overall for TPC-DS @ 100 GB.

tpcds_allqueries

**Comment by andygrove** [comment]
Average of 3 runs, main branch versus this PR.

tpcds_allqueries

Command used for both runs:

```shell
$SPARK_HOME/bin/spark-submit \
    --master $SPARK_MASTER \
    --conf spark.driver.memory=8G \
    --conf spark.executor.instances=1 \
    --conf spark.executor.memory=32G \
    --conf spark.executor.cores=8 \
    --conf spark.cores.max=8 \
    --conf spark.eventLog.enabled=true \
    --jars $COMET_JAR \
    --conf spark.driver.extraClassPath=$COMET_JAR \
    --conf spark.executor.extraClassPath=$COMET_JAR \
    --conf spark.sql.extensions=org.apache.comet.CometSparkSessionExtensions \
    --conf spark.comet.enabled=true \
    --conf spark.comet.exec.enabled=true \
    --conf spark.comet.exec.all.enabled=true \
    --conf spark.comet.cast.allowIncompatible=true \
    --conf spark.comet.shuffle.enforceMode.enabled=true \
    --conf spark.comet.exec.shuffle.enabled=true \
    --conf spark.comet.exec.shuffle.mode=auto \
    --conf spark.shuffle.manager=org.apache.spark.sql.comet.execution.shuffle.CometShuffleManager \
    tpcbench.py \
    --benchmark tpcds \
    --data /mnt/bigdata/tpcds/sf100/ \
    --queries ../../tpcds/queries-spark \
    --iterations 3
```

**Review by huaxingao — APPROVED** [review]
LGTM. Thanks for the PR @andygrove

File: native/core/src/execution/datafusion/planner.rs [hunk 1]
```diff
@@ -1234,24 +1233,51 @@ impl PhysicalPlanner {
     ) -> Result<Arc<dyn AggregateExpr>, ExecutionError> {
         match spark_expr.expr_struct.as_ref().unwrap() {
             AggExprStruct::Count(expr) => {
-                let children = expr
-                    .children
-                    .iter()
-                    .map(|child| self.create_expr(child, schema.clone()))
-                    .collect::<Result<Vec<_>, _>>()?;
+                if expr.children.iter().len() == 1 {
```
Comment (viirya) [inline native/core/src/execution/datafusion/planner.rs:1236]: Hmm, I think we can also do this for multiple child expressions?

**Review by viirya — APPROVED** [review]
Looks okay. Actually it is how Spark `count` does internally:

```
 /* count = */ If(nullableChildren.map(IsNull).reduce(Or), count, count + 1L)
```

File: native/core/src/execution/datafusion/planner.rs (hunk 1 above)
Comment (andygrove) [reply native/core/src/execution/datafusion/planner.rs]: Thanks. I have extended this approach for the multiple argument case.

**Pull request merged**